zebra-network/src/discoverer.rs

```diff
@@ -0,0 +1,166 @@
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::Stream;
+use pin_project::pin_project;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tower::{
+    discover::{Change, Discover},
+    load::{CompleteOnResponse, PeakEwma, PeakEwmaDiscover},
+    BoxError,
+};
+
+use crate::{
+    constants,
+    peer::{Client, LoadTrackedClient},
+    protocol::external::types::Version,
+};
+
+type LoadTracker = PeakEwmaDiscover<UnboundedReceiverStream<DiscoveryEvent<Client>>>;
+type DiscoveryEvent<C> = Result<Change<SocketAddr, C>, BoxError>;
+type PollDiscover = Poll<Option<DiscoveryEvent<LoadTrackedClient>>>;
+
+/// A collector of discovered peers that provides them as [`LoadTrackedClient`] services through a
+/// [`Discover`] interface.
+#[pin_project(project = PinnedPeerDiscoverer)]
+pub struct PeerDiscoverer<D> {
```

Comment:
Overall Module Question: This type is really complicated, I'm wondering if we could use a stream mapping instead?

It would look something like:
```rust
client_stream.map(|(address, client)| (
    address,
    LoadTrackedClient::new(PeakEwma::new(client, …), client.version),
))
```

This is the same wrapping that `PeakEwmaDiscover` does - there isn't any internal or cross-service state:
https://docs.rs/tower/0.4.11/src/tower/load/peak_ewma.rs.html#202

Changing this type would obsolete a bunch of other comments I've made, so I've marked them as resolved.